gradle-conjure/src/main/java/com/palantir/gradle/conjure/ConjureRunnerResource.java

```diff
@@ -37,70 +34,54 @@
 import net.bytebuddy.dynamic.loading.ClassLoadingStrategy;
 import net.bytebuddy.matcher.ElementMatchers;
 import net.bytebuddy.pool.TypePool;
-import org.gradle.api.Project;
-import org.gradle.api.file.RegularFileProperty;
-import org.gradle.api.services.BuildService;
-import org.gradle.api.services.BuildServiceParameters;
+import org.gradle.process.ExecOperations;
 import org.gradle.process.ExecResult;
+import org.gradle.workers.WorkQueue;
+import org.gradle.workers.WorkerExecutor;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-@SuppressWarnings("PublicConstructorForAbstractClass")
-public abstract class ConjureRunnerResource implements BuildService<Params>, Closeable {
+public abstract class ConjureRunnerResource {
 
     private static final Logger log = LoggerFactory.getLogger(ConjureRunnerResource.class);
 
-    public interface Params extends BuildServiceParameters {
+    interface ConjureRunner {
 
-        RegularFileProperty getExecutable();
+        void invoke(ExecOperations execOperations, String failedTo, List<String> unloggedArgs, List<String> loggedArgs);
     }
 
-    private final ConjureRunner delegate;
-
-    public ConjureRunnerResource() throws IOException {
-        this.delegate =
-                createNewRunner(getParameters().getExecutable().getAsFile().get());
-    }
-
-    final void invoke(Project project, String failedTo, List<String> unloggedArgs, List<String> loggedArgs) {
-        delegate.invoke(project, failedTo, unloggedArgs, loggedArgs);
-    }
+    static ConjureRunner createNewRunnerWithCurrentClasspath(File executable) throws IOException {
+        Optional<StartScriptInfo> maybeJava = ReverseEngineerJavaStartScript.maybeParseStartScript(executable.toPath());
 
-    @Override
-    public final void close() throws IOException {
-        delegate.close();
+        if (maybeJava.isPresent()) {
+            ReverseEngineerJavaStartScript.StartScriptInfo info = maybeJava.get();
+            Optional<Method> mainMethod =
+                    getMainMethod(info.mainClass(), Thread.currentThread().getContextClassLoader());
+            if (mainMethod.isPresent()) {
+                return new InProcessConjureRunner(executable, mainMethod.get());
+            }
+        }
+        return new ExternalProcessConjureRunner(executable);
     }
 
-    interface ConjureRunner extends Closeable {
-
-        void invoke(Project project, String failedTo, List<String> unloggedArgs, List<String> loggedArgs);
+    static WorkQueue getIsolatedConjureWorkQueue(WorkerExecutor workerExecutor, File executable) {
+        return workerExecutor.classLoaderIsolation(classLoaderWorkerSpec -> conjureExecutableClasspath(executable)
```

Comment:
I think we need to either write a test or at least manually check that this reuses classloaders for the same conjure executable.

Reply:
Hmm, I suspect they may not be:

https://github.com/gradle/gradle/blob/56445ef7f3b2117631f4afbbd9845b73ec723f82/subprojects/workers/src/main/java/org/gradle/workers/internal/IsolatedClassloaderWorker.java#L52

Reply:
Moved back to the daemon because it looks like that will allow you to re-use the underlying classloader: https://github.com/gradle/gradle/blob/ab01e7fc05e4d14186b0a706e0737fe0b20f458f/subprojects/workers/src/main/java/org/gradle/workers/internal/WorkerDaemonServer.java#L98

Reply:
That didn't actually work -- although gradle purports to reuse the underlying classloader, I think there's a "bug" where it discards the whole daemon object and recreates it.

Leaving this in the classpath isolation mode - it will still have to construct the classloader a bunch of times but at least the classpaths will be GC'd correctly